### PR TITLE
Redmine #7075 Store absolute time with process information

### PR DESCRIPTION
**Pull request by kacfengine**
<pre>
commit 203dd9a7328d4557b913d489e00e84d98a1ab77a
Author: Stefan Weil <sw@weilnetz.de>
Date:   Mon Mar 30 20:35:36 2015

    Redmine #7075 Store absolute time with process information
    
    Processes in CFEngine are identified by their OS process identifier (pid)
    and their start time.
    
    Up to now, the start time was relative to the boot time, so it was not
    unique across system restarts.
    
    Using an absolute start time assures that the combination of pid and
    start time is unique. This fixes bug #7075.
    
    Signed-off-by: Stefan Weil <sw@weilnetz.de>

commit fd5b4f7f91a264e1ddfa10cf4d376a75d6430dcf
Author: Stefan Weil <sw@weilnetz.de>
Date:   Mon Mar 30 20:27:08 2015

    Add new function GetBootTime
    
    The new function will be used for Linux in the following commit.
    
    The code was taken from the old implementation of GetUptimeSeconds
    which now uses GetBootTime.
    
    An implementation for Windows can be added later (when needed).
    
    Signed-off-by: Stefan Weil <sw@weilnetz.de>
</pre>

### CODE REVIEW

**Comment by kacfengine** [comment]
trigger build

File: libenv/sysinfo.c [hunk 1]
```diff
@@ -2666,103 +2666,119 @@ static void GetCPUInfo(EvalContext *ctx)
...
+    {
 #if defined(BOOT_TIME_WITH_SYSINFO)         // Most GNU, Linux platforms
 
-    struct sysinfo s;
-    if (sysinfo(&s) == 0)
-    {
-       // Don't return yet, sanity checking below
-       boot_time = now - s.uptime;
-    }
+        time_t now = time(NULL);
```
Comment (jimis) [inline libenv/sysinfo.c]: What if "now" changes as the system proceeds? For example battery-less embedded devices will start with unset time, then set it via NTP.

Comment (kacfengine) [inline libenv/sysinfo.c]: In that case we get the same issue as before the fix, boot time will be zero (or another constant value), and processes may get the same start time unintentionally. Only after the time is set do we get unique start times.

In theory you could defeat this mechanism by setting the time backwards, but you'd have to be *insanely* unlucky to both have the process start at the same second that the previous one started *and* get the same PID. This is practically impossible unless your intention is to trigger it and you try over and over. Certainly it is much more likely that the original issue described in https://dev.cfengine.com/issues/7075 will happen using the old method, than triggering what I just described with the new method.

Comment (jimis) [inline libenv/sysinfo.c]: You're right, but I'm not arguing this.

What I'm commenting about is that `boot_time` is *static*, which means that
a long running process will have the wrong time for all its lifetime.

Comment (kacfengine) [inline libenv/sysinfo.c]: Aah, now I see what I mean, well spotted. You're right, that may be a problem. I wonder how expensive it is to find out uptime every time. Hmm, need to think about that one...

Comment (kacfengine) [inline libenv/sysinfo.c]: ...to find out *boot time* every time...

**Comment by kacfengine** [comment]
@jimis brought up a very good point that with this fix you can render CFEngine unable to kill any stale agent simply by adjusting the system time one second forwards or backwards. The reason is that the new boot time depends on current time, and so if you change it, all processes will now appear foreign to CFEngine, and it will not attempt to kill them if they have gone stale. This is not good.

Damn, I had high hopes for this fix.

**Pull request closed without merging**